### PR TITLE
[onert] Support TFlite 2.8 for gpu_cl

### PR DESCRIPTION
**Pull request by YiHyunJin**
1. Update the tensorflow version used in gpu_cl to 2.8.
2. I added config file to download source code to build tensorflow lite 2.8.
3. Applied changed api with update to 2.8.
4. Delete unnecessary code.

ONE-DCO-1.0-Signed-off-by: HyunJin Yi <hj0412.yi@samsung.com>

### CODE REVIEW

File: infra/cmake/packages/Egl_HeadersSourceConfig.cmake [hunk 1]
```diff
@@ -0,0 +1,21 @@
+function(_Egl_HeadersSource_import)
```
Comment (seanshpark) [inline infra/cmake/packages/Egl_HeadersSourceConfig.cmake:1]: please split infra in separate PR

**Comment by seanshpark** [comment]
Our current TF is 2.8.0. Is there particular reason to use TF 2.7.0 ?

File: infra/cmake/packages/Egl_HeadersSourceConfig.cmake (hunk 1 above)
Comment (YiHyunJin) [reply infra/cmake/packages/Egl_HeadersSourceConfig.cmake:1]: Yes, I will split it.

**Comment by YiHyunJin** [comment]
@seanshpark 
There's no reason to use 2.7 instead of 2.8. I updated to 2.7 because my existing working repo used 2.7.
Upgrading to 2.8 requires some modifications.

File: infra/cmake/packages/Egl_HeadersSourceConfig.cmake (hunk 1 above)
Comment (YiHyunJin) [reply infra/cmake/packages/Egl_HeadersSourceConfig.cmake:1]: I splited through (https://github.com/Samsung/ONE/pull/9257) . please check it. If the it is merged, i will proceed with the correction of this PR.

**Comment by seanshpark** [comment]
> There's no reason to use 2.7 instead of 2.8. I updated to 2.7 because my existing working repo used 2.7.
Upgrading to 2.8 requires some modifications.

**Comment by YiHyunJin** [comment]
> > There's no reason to use 2.7 instead of 2.8. I updated to 2.7 because my existing working repo used 2.7.
> > Upgrading to 2.8 requires some modifications.
> 
> Can you please share background or purpose of this change? If there is an issue for this, please add a link.

https://github.sec.samsung.net/tizen-sys-con/ONE/pull/185
https://github.sec.samsung.net/tizen-sys-con/ONE/pull/197

The OP not implemented in tf2.4 implemented ex separately.
However, it is implemented in tf2.7, so I updated it to tf2.7.
I don't want to use ex anymore.

**Comment by YiHyunJin** [comment]
@nnfw-bot test cross-runtime-test-debug
@nnfw-bot test cross-runtime-test-release

File: infra/nnfw/cmake/options/options_armv7l-tizen.cmake [hunk 2]
```diff
@@ -11,6 +11,5 @@ option(GENERATE_RUNTIME_NNAPI_TESTS "Generate NNAPI operation gtest" OFF)
 option(ENVVAR_ONERT_CONFIG "Use environment variable for onert configuration" OFF)
 
 option(DOWNLOAD_OPENCL_HEADERS "Download Opencl_headers source" ON)
-option(DOWNLOAD_TENSORFLOW_GPU "Download Tensorflow GPU delegate source" ON)
-option(BUILD_GPU_CL "Build gpu_cl backend" ON)
+option(BUILD_GPU_CL "Build gpu_cl backend" OFF)
```
Comment (chunseoklee) [inline infra/nnfw/cmake/options/options_armv7l-tizen.cmake:14]: @YiHyunJin Why you turn off build from tizen platform ?

Comment (YiHyunJin) [reply infra/nnfw/cmake/options/options_armv7l-tizen.cmake:14]: I turned it off to confirm. If you compress and upload the TF2.7 code now, it will not upload if the size is over 50MB.
So, I am trying to link tf2.7 directly through nnfw.spec.

Comment (YiHyunJin) [reply infra/nnfw/cmake/options/options_armv7l-tizen.cmake:14]: I turned it off to confirm. I compress and try to upload the TF2.7 code. But it did not upload. because the size is over 50MB.
So, I am trying to link tf2.7 directly through nsfw.spec.

Comment (chunseoklee) [reply infra/nnfw/cmake/options/options_armv7l-tizen.cmake:14]: What do you mean by "link tf2.7 directly through nnfw.spec" ?

Comment (YiHyunJin) [reply infra/nnfw/cmake/options/options_armv7l-tizen.cmake:14]: I think.. we need tf-lite 2.7 source code or lib.
I try to upload tf-lite 2.7 souce code, but It was failed..
So I will try to link "tensorflow-lite-devel" version 2.7
Is this solution incorrect??...

Comment (chunseoklee) [reply infra/nnfw/cmake/options/options_armv7l-tizen.cmake:14]: Have you check https://review.tizen.org/gerrit/#/admin/projects/platform/upstream/tensorflow2 is able to provide such package ? IMHO, I do not think so. Or you should split compression.

Comment (YiHyunJin) [reply infra/nnfw/cmake/options/options_armv7l-tizen.cmake:14]: `Or you should split compression.`
oh it is good idea. : )
Thanks. I will try to split compression

File: infra/nnfw/cmake/options/options_armv7l-linux.cmake [hunk 3]
```diff
@@ -4,7 +4,9 @@
 option(DOWNLOAD_NEON2SSE "Download NEON2SSE library source" OFF)
 option(BUILD_OPENCL_TOOL "Build OpenCL tool" ON)
 
+option(DOWNLOAD_OPENGL_HEADERS "Download Opengl_headers source" ON)
+option(DOWNLOAD_EGL_HEADERS "Download Egl_headers source" ON)
+option(DOWNLOAD_VULKAN "Download vulkan source" ON)
```
Comment (seanshpark) [inline infra/nnfw/cmake/options/options_armv7l-linux.cmake:9]: why are theses needed for linux?

Comment (YiHyunJin) [reply infra/nnfw/cmake/options/options_armv7l-linux.cmake:9]: These are needed for building tensorflow lite. So I added it.

**Comment by YiHyunJin** [comment]
> > https://github.sec.samsung.net/tizen-sys-con/ONE/pull/185
> > https://github.sec.samsung.net/tizen-sys-con/ONE/pull/197
> 
> Thesw two are PRs.
> 
> I asked `Can you please share background or purpose of this change? If there is an issue for this, please add a link.`
> 
> I don't understand how I could get the background or purpose of this PR.

```
The OP not implemented in tf2.4 implemented ex separately.
However, it is implemented in tf2.7, so I updated it to tf2.7.
I don't want to use ex anymore.
```

Ex folder was removed in these PRs.
So I linked these PRs.

We need Reduce Op, Split Op.... But  Those OPs are not in TF2.4.1, but are in TF2.7.0.
So I update TF2.4.1 to TF2.7.0.

**Comment by YiHyunJin** [comment]
> > Ex folder was removed in these PRs.
> > So I linked these PRs.
> > We need Reduce Op, Split Op.... But Those OPs are not in TF2.4.1, but are in TF2.7.0.
> > So I update TF2.4.1 to TF2.7.0.
> 
> Are these only for Tizen? I know you work on Tizen so I can understand the changes for Tizen. I see changes for Android and also for Linux? But why these too? There are too many changes that just `So I update TF2.4.1 to TF2.7.0.` explanation is not enough. I don't review codes related for OneRT so actully don't care, but there are changes related with common part in `infra` (so I requested to separate them) and I also sometimes test with OneRT as a user and `options_armv7l-linux.cmake` changes affect this.


I am currently developing gpu_cl. I am updating the tensorflow version used in gpu_cl to 2.7 now. 
For this reason, 
`We need Reduce Op, Split Op.... But Those OPs are not in TF2.4.1, but are in TF2.7.0.`
however, the gpu_cl test is running on Tizen, Andorid, and Linux. So I fixed them all.
Changing the Tensorflow version will change the api we use. So it is difficult to change only the tizen.

**Comment by seanshpark** [comment]
> however, the gpu_cl test is running on Tizen, Andorid, and Linux. So I fixed them all.

Does this change work also in RPi4 Ubuntu 18.04/20.04?

**Comment by YiHyunJin** [comment]
> Does this change work also in RPi4 Ubuntu 18.04/20.04?

I didn't check it directly, but I checked it through "cross-runtime-test-debug" and "cross-runtime-test-release".

File: infra/nnfw/cmake/options/options_armv7l-linux.cmake (hunk 3 above)
Comment (seanshpark) [reply infra/nnfw/cmake/options/options_armv7l-linux.cmake:9]: As I understand these are options and these can be swiched on from the ourside.
Settings these ON is these are necessary for the build and turning it off may break the build.

Are these options necessary for OneRT build for linux?
If not please turn it off and switch ON from outside in your build environment.

File: infra/nnfw/cmake/options/options_armv7l-tizen.cmake (hunk 2 above)
Comment (chunseoklee) [reply infra/nnfw/cmake/options/options_armv7l-tizen.cmake:14]: @YiHyunJin Yes.

File: infra/nnfw/cmake/options/options_armv7l-linux.cmake (hunk 3 above)
Comment (chunseoklee) [reply infra/nnfw/cmake/options/options_armv7l-linux.cmake:9]: @seanshpark I have checked again the gpu_cl backend history about FLAG in [infra/nnfw/cmake/options/options_armv7l-linux.cmake](https://github.com/Samsung/ONE/pull/9256/files/9c58d6b68222d3c968cbe84fe39b48d04d24fa00#diff-88596ec4ef6912b5d8f9d29c5465de1ec40528a85e1db095415d208e6f3a7b0b)

1. nnfw_api test(unit test) about gpu_cl backend is running on odroid-xu4 in CI server. Thus, the following option should be on default

```
option(BUILD_GPU_CL "Build gpu_cl backend" ON)
option(BUILD_TENSORFLOW_LITE_GPU "Build TensorFlow Lite GPU delegate from the downloaded source" ON)
```

Thus, the patch below might be necessary for the build with tensorflow 2.7 source : 

```
option(DOWNLOAD_OPENGL_HEADERS "Download Opengl_headers source" ON)
option(DOWNLOAD_EGL_HEADERS "Download Egl_headers source" ON)
option(DOWNLOAD_VULKAN "Download vulkan source" ON)
```

IMHO, to update into tf2.7 with gpu_gl, this option is mandatory.

Sorry for confusing.

**Comment by lemmaa** [comment]
@YiHyunJin , can you rebase this PR to version 2.8 of TensorFlow and send it back if it doesn't require a lot of effort?

ONE's TensorFlow support history is a bit complicated. It has been proceeding in the order of 2.2 -> 2.3 -> 2.4 -> 2.6 -> 2.8. Currently, version 2.8 is officially supported, but some modules are still stuck in the old version. For version 2.7, we never supported it. It doesn't seem appropriate to introduce new kernels to a version that has never been supported in the past.

Sooner or later, delayed upgrades of several modules will all have to match 2.8. In addition, it is necessary to upgrade again to version 2.9 afterward. 

Considering these various situations, it is recommended to upgrade your work to support version 2.8 immediately.

**Comment by YiHyunJin** [comment]
> @YiHyunJin , how about rebase this PR to version 2.8 of TensorFlow and send it back if it doesn't require a lot of effort?
> 
> ONE's TensorFlow support history is a bit complicated. It has been proceeding in the order of 2.2 -> 2.3 -> 2.4 -> 2.6 -> 2.8. Currently, version 2.8 is officially supported, but some modules are still stuck in the old version. For version 2.7, we never supported it. It doesn't seem appropriate to introduce new kernels to a version that has never been supported in the past.
> 
> Sooner or later, delayed upgrades of several modules will all have to match 2.8. In addition, it is necessary to upgrade again to version 2.9 afterward.
> 
> Considering these various situations, it is recommended to upgrade your work to support version 2.8 immediately.

Okay. I will rebase this PR to version 2.8. :)

**Comment by YiHyunJin** [comment]
@nnfw-bot cross-aarch64-runtime-test-release

**Review by intom — COMMENTED** [review]
Refer to the below diff for error
```
diff --git a/infra/nnfw/cmake/packages/TensorFlowGpuConfig.cmake b/infra/nnfw/cmake/packages/TensorFlowGpuConfig.cmake
index 7c8288d7b..b27aa362e 100644
--- a/infra/nnfw/cmake/packages/TensorFlowGpuConfig.cmake
+++ b/infra/nnfw/cmake/packages/TensorFlowGpuConfig.cmake
@@ -1,4 +1,11 @@
 # TensorFlowGpuConfig.cmake
+macro(return_unless VAR)
+if(NOT ${VAR})
+  message("TensorFlowGpu: ${VAR} NOT TRUE")
+  set(TensorFlowGpu_FOUND FALSE PARENT_SCOPE)
+  return()
+endif(NOT ${VAR})
+endmacro(return_unless)
 
 function(_Build_TfliteGpuDelagate_)
   nnas_find_package(TensorFlowSource EXACT 2.8.0 QUIET)
@@ -16,7 +23,8 @@ function(_Build_TfliteGpuDelagate_)
   nnas_find_package(Farmhash REQUIRED)
   return_unless(Farmhash_FOUND)
 
-  nnas_find_package(FlatBuffers REQUIRED)
+  # nnas_find_package(FlatBuffers REQUIRED)
+  # return_unless(FlatBuffers_FOUND)
 
   nnas_find_package(Fp16Source REQUIRED)
   return_unless(Fp16Source_FOUND)
diff --git a/runtime/onert/backend/gpu_cl/CMakeLists.txt b/runtime/onert/backend/gpu_cl/CMakeLists.txt
index 4b7cae44e..6fd1adeb6 100644
--- a/runtime/onert/backend/gpu_cl/CMakeLists.txt
+++ b/runtime/onert/backend/gpu_cl/CMakeLists.txt
@@ -59,6 +59,9 @@ target_compile_options(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE "-DCL_TARGET_OPENCL_V
 
 add_library(tflite_ignore_warnings INTERFACE)
 target_compile_options(tflite_ignore_warnings INTERFACE -Wno-unused-parameter -Wno-sign-compare)
+if (${TARGET_OS} MATCHES "tizen")
+    target_compile_options(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE "-Wno-error=deprecated-copy")
+endif ()
 target_link_libraries(${LIB_ONERT_BACKEND_GPU_CL} INTERFACE tflite_ignore_warnings)
 
 target_link_libraries(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE abseil)
@@ -71,7 +74,7 @@ target_link_libraries(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE onert_core)
 target_link_libraries(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE ${LIB_ONERT_BACKEND_CL_COMMON})
 target_link_libraries(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE nnfw_common)
 target_link_libraries(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE nnfw_coverage)
-target_link_libraries(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE flatbuffers::flatbuffers)
+target_link_libraries(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE flatbuffers)
 
 set_target_properties(${LIB_ONERT_BACKEND_GPU_CL} PROPERTIES OUTPUT_NAME backend_gpu_cl)
```

**Comment by YiHyunJin** [comment]
PTAL :)

File: infra/nnfw/cmake/packages/TensorFlowLiteGpu/CMakeLists.txt [hunk 4]
```diff
@@ -1,43 +1,32 @@
 #
-# Tensorflow Lite GPU delegate library 2.4.1
+# Tensorflow Lite GPU delegate library 2.7.0
```
Comment (intom) [inline infra/nnfw/cmake/packages/TensorFlowLiteGpu/CMakeLists.txt:2]: is TensorFlowLiteGpu on 2.7.0?

Comment (YiHyunJin) [reply infra/nnfw/cmake/packages/TensorFlowLiteGpu/CMakeLists.txt:2]: Thanks. : )

**Comment by YiHyunJin** [comment]
PTAL :)

File: runtime/onert/backend/gpu_cl/CMakeLists.txt [hunk 5]
```diff
@@ -35,18 +54,36 @@ file(GLOB_RECURSE SOURCES "*.cc")
...
+target_include_directories(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE ${Egl_HeadersSource_DIR}/api)
+
+if (${TARGET_OS} MATCHES "tizen")
+    target_compile_options(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE "-Wno-error=deprecated-copy")
+endif ()
+
+target_compile_options(${LIB_ONERT_BACKEND_GPU_CL} PRIVATE "-DCL_TARGET_OPENCL_VERSION=220" "-DEGL_NO_X11")
+
+add_library(tflite_ignore_warnings INTERFACE)
+target_compile_options(tflite_ignore_warnings INTERFACE -Wno-unused-parameter -Wno-sign-compare)
```
Comment (chunseoklee) [inline runtime/onert/backend/gpu_cl/CMakeLists.txt:69]: duplicated with 'gcc diagnostic' preprocessor ?

**Review by chunseoklee — CHANGES_REQUESTED** [review]

File: runtime/onert/backend/gpu_cl/Utils.h [hunk 6]
```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
```
Comment (chunseoklee) [inline runtime/onert/backend/gpu_cl/Utils.h:2]: 2022

**Review by chunseoklee — CHANGES_REQUESTED** [review]

File: runtime/onert/backend/gpu_cl/CMakeLists.txt (hunk 5 above)
Comment (YiHyunJin) [reply runtime/onert/backend/gpu_cl/CMakeLists.txt]: I deleted it. Thanks :)

**Comment by seanshpark** [comment]
title `Supoort` typo

File: runtime/onert/backend/gpu_cl/CMakeLists.txt (hunk 5 above)
Comment (YiHyunJin) [reply runtime/onert/backend/gpu_cl/CMakeLists.txt]: I will check

**Comment by ragmani** [comment]
@nnfw-bot test cross-runtime-test-debug

File: runtime/onert/backend/gpu_cl/CMakeLists.txt (hunk 5 above)
Comment (chunseoklee) [reply runtime/onert/backend/gpu_cl/CMakeLists.txt]: @YiHyunJin You got me wrong. My intention is removing gcc preprocessor for tidy code

Comment (YiHyunJin) [reply runtime/onert/backend/gpu_cl/CMakeLists.txt]: @chunseoklee ah... okay. i understand, i will check and modify.

File: infra/nnfw/cmake/options/options_armv7l-linux.cmake (hunk 3 above)
Comment (chunseoklee) [reply infra/nnfw/cmake/options/options_armv7l-linux.cmake:9]: @YiHyunJin Have you seen this error log ? 

```
ubuntu@ubuntu ~> Product/out/bin/nnpackage_run mobilenet_v1_1.0_224_quant/
Package Filename mobilenet_v1_1.0_224_quant/
Product/out/bin/nnpackage_run: symbol lookup error: /home/ubuntu/Product/out/bin/../lib/../lib/../lib/libbackend_gpu_cl.so: undefined symbol: _ZN4absl12lts_2021110212UnknownErrorENS0_11string_viewE
```

I have seen this on rpi4 armv7l ubuntu. ( note that rpi does not provide OpenCL DDK)

**Review by hseok-oh — APPROVED** [review]
I have small question on cmake implementation. But we delay this PR too long. Let's merge this first then I'll look more carefully.

**Review by hseok-oh — DISMISSED** [review]

**Review by hseok-oh — APPROVED** [review]

**Review by chunseoklee — APPROVED** [review]

**Comment by seanshpark** [comment]
Please go on without my approval.

**Comment by hseok-oh** [comment]
I removed @glistening from reviewer and merged because there is no reaction for a long time.
